Source/AssetRipper.Processing/AnimationClips/CustomCurveResolver.cs

```diff
@@ -20,505 +19,328 @@ public CustomCurveResolver(IAnimationClip clip)
 
 		public string ToAttributeName(BindingCustomType type, uint attribute, string path)
 		{
+			int i;
+			string? foundPath;
+
 			switch (type)
 			{
 				case BindingCustomType.BlendShape:
 					{
 						const string Prefix = "blendShape.";
+
 						if (UnknownPathRegex().IsMatch(path))
 						{
 							return Prefix + attribute;
 						}
 
-						foreach (IGameObject root in Roots)
+						for (i = 0; i < Roots.Length; i++)
 						{
-							ITransform rootTransform = root.GetTransform();
-							ITransform? child = rootTransform.FindChild(path);
-							if (child == null)
+							if (!Roots[i].TryGetChildComponent(path, out ISkinnedMeshRenderer? skin))
 							{
 								continue;
 							}
-							ISkinnedMeshRenderer? skin = child.GetGameObject().TryGetComponent<ISkinnedMeshRenderer>();
-							if (skin == null)
-							{
-								continue;
-							}
-							IMesh? mesh = skin.Mesh_C137P;
-							if (mesh == null)
+
+							if (skin.Mesh_C137P is not IMesh mesh)
 							{
 								continue;
 							}
-							string? shapeName = mesh.FindBlendShapeNameByCRC(attribute);
-							if (shapeName == null)
+
+							if (mesh.FindBlendShapeNameByCRC(attribute) is not string shapeName)
 							{
 								continue;
 							}
 
 							return Prefix + shapeName;
 						}
+
 						return Prefix + attribute;
 					}
 
 				case BindingCustomType.Renderer:
-					return "m_Materials"
-						+ "." + "Array" //from the common string
-						+ "." + "data" //from the common string
-						+ $"[{attribute}]";
+					return $"m_Materials.Array.data[{attribute}]"; //from the common string
 
 				case BindingCustomType.RendererMaterial:
 					{
 						const string Prefix = "material.";
+
 						if (UnknownPathRegex().IsMatch(path))
 						{
 							return Prefix + attribute;
 						}
 
-						foreach (IGameObject root in Roots)
+						uint crc28 = attribute & 0xFFFFFFF;
+						bool usesRgba = (attribute & (1 << 30)) != 0;
+						bool hasNoSubProperty = (attribute & (1 << 31)) != 0;
+						uint subPropetyIndex = (attribute >> 28) & 3;
+
+						for (i = 0; i < Roots.Length; i++)
 						{
-							ITransform rootTransform = root.GetTransform();
-							ITransform? child = rootTransform.FindChild(path);
-							if (child == null)
+							if (!Roots[i].TryGetChildComponent(path, out IRenderer? renderer))
 							{
 								continue;
 							}
 
-							uint crc28 = attribute & 0xFFFFFFF;
-							IRenderer? renderer = child.GetGameObject().TryGetComponent<IRenderer>();
-							if (renderer == null)
-							{
-								continue;
-							}
-							string? property = renderer.FindMaterialPropertyNameByCRC28(crc28);
-							if (property == null)
+							if (renderer.FindMaterialPropertyNameByCRC28(crc28) is not string property)
```

Comment:
My understanding is that your intention is to do a null check here rather than a type check. Please change this to `is not { } property`.